File: src/algorithm/leetcode/Problem560.java
package algorithm.leetcode;

import java.util.HashMap;

public class Problem560 {

    public static int subarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> hashMap = new HashMap<>();
        hashMap.put(0, 1);
        int sum = 0;
        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (hashMap.containsKey(sum - k)) {
                count += hashMap.get(sum - k);
            }
            hashMap.putIfAbsent(sum, 0);
            hashMap.put(sum, hashMap.get(sum) + 1);
        }
        return count;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1,1,1};
        System.out.println(subarraySum(nums, 2));
    }

}

File: src/algorithm_cpp/problem560.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
	map<int, int> precount;
	int count = 0, pre = 0;
	precount[0] = 1;
	for (int i = 0; i < nums.size(); i++) {
	    pre += nums[i];
	    if (precount.find(pre - k) != precount.end()) {
		count += precount[pre - k];
	    }
	    precount[pre] += 1;
	}
	return count;
    }
};
